### ISSUE TITLE
race condition in JSON format

### ISSUE THREAD

**Issue by yawnt** [unicredit/sbt-swagger-codegen#4]
sometimes something like this happens

generated models:

``` scala
case class X(param: Y)
case class Y(a: String)
```

generated json serializers

``` scala
val first = Json.format[X]
val second = Json.format[Y]
```

which fails at the first line because there is no formatter for Y in scope yet

**Comment by andreaTP** [comment]
There is the need of adding a topological sorting to formatter order, this will work for **acyclic** dependencies

**Comment by fralken** [comment]
Another option could be not using macros. In this case it would work also for cyclic dependencies.

**Comment by andreaTP** [comment]
if you have a POC on this I will go for this option @fralken  :-)

**Comment by fralken** [comment]
I implemented in fa5f138 the generation of implicit `Reads` and `Writes` without using macros. This solves the issue of the forward references.
Now `Reads` and `Writes` are generated in the same way as the macro would do. This works only for case classes with no more than 22 parameters (due to a limitation in `play.api.libs.functional` of `FunctionalBuilder`). For case classes with more than 22 parameters e different and more explicit code is generated.

**Comment by andreaTP** [comment]
We can close this
